librubyfmt/src/format_prism.rs

```diff
@@ -235,6 +242,184 @@ fn format_statements(ps: &mut dyn ConcreteParserState, statements_node: prism::S
+
+    // `opening_loc()` is only `None` in the case of the inner parts of multiline strings
+    // (e.g. the inner contents of a heredoc)
+    let opener = string_node
+        .opening_loc()
+        .map(|s| loc_to_string(s).trim().to_string());
+    let closer = string_node
+        .closing_loc()
+        .map(|s| loc_to_string(s).trim().to_string());
+    let is_heredoc = opener.clone().map(|s| s.starts_with("<")).unwrap_or(false);
+
+    ps.with_start_of_line(
+        false,
+        Box::new(|ps| {
+            if is_heredoc {
+                let heredoc_symbol = opener.clone().unwrap();
+                let heredoc_kind = HeredocKind::from_string(&heredoc_symbol);
+
+                ps.emit_heredoc_start(heredoc_symbol, heredoc_kind);
+                ps.emit_newline();
+            } else {
+                // Always use double quotes over single quotes/percent literals
+                if opener.is_some() {
+                    ps.emit_double_quote();
+                }
+            }
+
+            // If opener is nil, we must be in some kind of interpolated string context, which
+            // means the contents must already be appropriately escaped -- hence we default to `true` here
+            let in_escaped_context =
+                is_heredoc || opener.clone().map(|s| s.starts_with("\"")).unwrap_or(true);
+            let string_content = if in_escaped_context {
+                loc_to_string(string_node.content_loc())
+            } else {
+                crate::string_escape::single_to_double_quoted(
+                    loc_to_string(string_node.content_loc()),
+                    opener.clone().unwrap().as_str(),
+                    closer.clone().unwrap().as_str(),
+                )
+            };
+
+            ps.emit_string_content(string_content);
+            ps.wind_dumping_comments_until_offset(string_node.content_loc().end_offset());
+
+            if is_heredoc {
+                // <<~ heredocs have their closing tag indented too
+                if opener.map(|s| s.starts_with("<<~")).unwrap_or(false) {
+                    ps.emit_indent();
+                }
+
+                ps.emit_heredoc_close(
+                    loc_to_string(string_node.closing_loc().unwrap())
+                        .trim()
+                        .to_string(),
+                );
+            } else if opener.is_some() {
+                ps.emit_double_quote();
+            }
+        }),
+    );
+
+    ps.wind_dumping_comments_until_offset(string_node.location().end_offset());
+
+    if ps.at_start_of_line() {
+        ps.emit_newline();
+    }
+}
+
+fn format_interpolated_string_node(
+    ps: &mut dyn ConcreteParserState,
+    interpolated_string_node: prism::InterpolatedStringNode,
+) {
+    if ps.at_start_of_line() {
+        ps.emit_indent();
+    }
+
+    let is_heredoc = interpolated_string_node
+        .opening_loc()
+        .map(|s| loc_to_string(s).starts_with("<"))
```

Comment:
`loc_to_str` at some point might be nice, also for the `loc_to_string.trim.to_string` patterns in this file.